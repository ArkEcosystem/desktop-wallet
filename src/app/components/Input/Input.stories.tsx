import React from "react";
import { Input, InputPassword } from "./index";

export default {
	title: "Components / Input",
};

export const Default = () => {
	return (
		<div className="max-w-xs space-y-4">
			<Input type="text" placeholder="Enabled" />
			<Input type="text" placeholder="Invalid" aria-invalid={true} />
			<Input type="text" placeholder="Disabled" disabled />
		</div>
	);
};

export const Password = () => {
	return (
		<div className="max-w-xs">
			<InputPassword defaultValue="secret" />
		</div>
	);
};
