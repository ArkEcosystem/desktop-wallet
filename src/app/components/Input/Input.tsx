import React from "react";
import tw, { styled } from "twin.macro";

import { useFormField } from "../Form/useFormField";

type InputProps = { as?: React.ElementType; isInvalid?: boolean } & React.HTMLProps<any>;

const InputStyled = styled.input`
	&:focus {
		${tw`border-2 outline-none border-theme-primary`}
	}
	&::placeholder {
		${tw`text-theme-neutral-light`}
	}
	&:disabled {
		${tw`border-theme-neutral-300 bg-theme-neutral-contrast text-theme-neutral-dark`}
	}
	&[aria-invalid="true"] {
		${tw`border-theme-danger`}
	}
`;

type InputElement = HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement;

export const Input = React.forwardRef<InputElement, InputProps>(
	({ isInvalid, className, ...props }: InputProps, ref) => {
		const fieldContext = useFormField();

		return (
			<InputStyled
				data-testid="Input"
				className={`overflow-hidden w-full bg-theme-background appearance-none rounded border border-theme-neutral-300 text-theme-neutral-900 transition-colors duration-200 px-4 py-3 ${
					className || ""
				}`}
				name={fieldContext?.name}
				aria-invalid={fieldContext?.isInvalid || isInvalid}
				ref={ref}
				{...props}
			/>
		);
	},
);

Input.displayName = "Input";
Input.defaultProps = {};
