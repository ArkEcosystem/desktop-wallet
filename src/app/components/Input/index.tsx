export * from "./Input";
export * from "./InputGroup";
export * from "./InputPassword";
export * from "./InputCounter";
export * from "./InputRange";
