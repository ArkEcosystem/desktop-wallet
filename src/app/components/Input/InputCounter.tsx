import React, { useState } from "react";

import { useFormField } from "../Form/useFormField";
import { Input } from "./Input";
import { InputAddonEnd, InputGroup } from "./InputGroup";

type Props = {
	maxLength?: number;
	maxLengthLabel?: string;
	defaultValue?: string;
	errorClassName?: string;
} & React.InputHTMLAttributes<any>;

export const InputCounter = React.forwardRef<HTMLInputElement, Props>((props: Props, ref) => {
	const fieldContext = useFormField();
	const [length, setLength] = useState(props.defaultValue?.length || 0);

	const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		setLength(event.target.value.length);
		props.onChange?.(event);
	};

	return (
		<InputGroup>
			<Input
				data-testid="InputCounter__input"
				className={fieldContext?.isInvalid ? "pr-28" : "pr-18"}
				ref={ref}
				{...props}
				onChange={handleChange}
			/>
			<InputAddonEnd
				data-testid="InputCounter__counter"
				className={`pl-2 text-sm pointer-events-none text-theme-secondary-500 ${
					fieldContext?.isInvalid ? "pr-12" : "pr-2"
				}`}
			>
				{length}/{props.maxLengthLabel}
			</InputAddonEnd>
		</InputGroup>
	);
});

InputCounter.displayName = "InputCounter";
