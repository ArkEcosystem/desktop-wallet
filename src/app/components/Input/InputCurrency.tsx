import { Currency } from "@arkecosystem/platform-sdk-intl";
import React from "react";

import { Input } from "./Input";

type Props = { onChange?: (value: string) => void; magnitude?: number } & Omit<
	React.InputHTMLAttributes<any>,
	"onChange" | "defaultValue"
>;

export const InputCurrency = React.forwardRef<HTMLInputElement, Props>(
	({ onChange, value, magnitude, ...props }: Props, ref) => {
		const convertValue = React.useCallback((value: string) => Currency.fromString(value, magnitude), [magnitude]);
		const defaultValue = value?.toString() || "";
		const [amount, setAmount] = React.useState(convertValue(defaultValue));

		React.useLayoutEffect(() => {
			setAmount(convertValue(defaultValue));
		}, [defaultValue, convertValue]);

		const handleInput = (event: React.ChangeEvent<HTMLInputElement>) => {
			const currency = convertValue(event.target.value);
			setAmount(currency);
			onChange?.(currency.value);
		};

		return (
			<Input
				data-testid="InputCurrency"
				type="text"
				value={amount.display}
				onChange={handleInput}
				ref={ref}
				{...props}
			/>
		);
	},
);

InputCurrency.displayName = "InputCurrency";
InputCurrency.defaultProps = {
	magnitude: 8,
};
