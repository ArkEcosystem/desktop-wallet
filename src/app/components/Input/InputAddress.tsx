import { Coins } from "@arkecosystem/platform-sdk";
import { useFormField } from "app/components/Form/useFormField";
import { Icon } from "app/components/Icon";
import { useEnvironmentContext } from "app/contexts";
import cn from "classnames";
import React from "react";
import { ValidationRules } from "react-hook-form";
import { useTranslation } from "react-i18next";

import { Input } from "./Input";
import { InputAddonEnd, InputGroup } from "./InputGroup";

export type InputAddressProps = {
	coin?: string;
	network?: string;
	registerRef?: (options: ValidationRules) => (ref: HTMLInputElement | null) => void;
	additionalRules?: ValidationRules;
	onValidAddress?: (address: string) => void;
	onQRCodeClick?: () => void;
	onChange?: (address: string) => void;
	useDefaultRules?: boolean;
} & React.InputHTMLAttributes<any>;

export const InputAddress = ({
	coin,
	network,
	registerRef,
	additionalRules,
	onValidAddress,
	onQRCodeClick,
	useDefaultRules,
	...props
}: InputAddressProps) => {
	const { t } = useTranslation();
	const { env } = useEnvironmentContext();
	const fieldContext = useFormField();

	const validateAddress = async (address: string) => {
		const instance: Coins.Coin = await env.coin(coin!, network!);
		const isValidAddress: boolean = await instance.identity().address().validate(address);

		if (isValidAddress) {
			onValidAddress?.(address);
			return true;
		}

		return t("COMMON.INPUT_ADDRESS.VALIDATION.NOT_VALID");
	};

	const defaultRules = {
		...additionalRules,
		validate: {
			validateAddress,
			...additionalRules?.validate,
		},
	};
	const rules = useDefaultRules ? defaultRules : {};

	return (
		<InputGroup className="max-w-20">
			<Input
				ref={registerRef?.(rules)}
				type="text"
				className={cn({ "pr-12": !fieldContext?.isInvalid, "pr-20": fieldContext?.isInvalid })}
				errorClassName="mr-12"
				data-testid="InputAddress__input"
				{...props}
			/>
			<InputAddonEnd className="my-px mr-4">
				<button
					data-testid="InputAddress__qr-button"
					type="button"
					className="flex justify-center items-center w-full h-full text-2xl focus:outline-none text-theme-primary-400"
					onClick={onQRCodeClick}
				>
					<Icon name="QrCode" width={20} height={20} />
				</button>
			</InputAddonEnd>
		</InputGroup>
	);
};

InputAddress.defaultProps = {
	additionalRules: {},
	useDefaultRules: true,
};
