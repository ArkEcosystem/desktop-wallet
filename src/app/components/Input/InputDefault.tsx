import { Input } from "app/components/Input";
import { InputGroup } from "app/components/Input";
import cn from "classnames";
import React from "react";

import { useFormField } from "../Form/useFormField";

type InputDefaultProps = {
	className?: string;
} & React.InputHTMLAttributes<any>;

export const InputDefault = React.forwardRef<HTMLInputElement, InputDefaultProps>(
	({ className, ...props }: InputDefaultProps, ref) => {
		const fieldContext = useFormField();

		return (
			<InputGroup>
				<Input className={cn(className, { "pr-12": fieldContext?.isInvalid })} ref={ref} {...props} />
			</InputGroup>
		);
	},
);

InputDefault.displayName = "InputDefault";
