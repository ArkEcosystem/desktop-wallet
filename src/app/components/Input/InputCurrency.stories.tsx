import { action } from "@storybook/addon-actions";
import React from "react";

import { InputCurrency } from "./InputCurrency";

export default { title: "Input / Input / InputCurrency" };

export const Default = () => {
	return <InputCurrency onChange={action("onChange")} />;
};
