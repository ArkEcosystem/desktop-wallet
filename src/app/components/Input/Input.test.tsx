import React from "react";
import { render } from "test-utils";

import { Input } from "./Input";

describe("Input", () => {
	it("should render a default input", () => {
		const { asFragment } = render(<Input />);

		expect(asFragment()).toMatchSnapshot();
	});

	it("should render with a custom element", () => {
		const { getByTestId } = render(<Input as="select" />);
		const input = getByTestId("Input");
		expect(input.tagName).toEqual("SELECT");
	});
});
