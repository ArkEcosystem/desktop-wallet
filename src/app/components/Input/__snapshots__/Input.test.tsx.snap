// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render a default input 1`] = `
<DocumentFragment>
  <input
    class="sc-AxjAm jCGFzm overflow-hidden w-full bg-theme-background appearance-none rounded border border-theme-neutral-300 text-theme-neutral-900 transition-colors duration-200 px-4 py-3 "
    data-testid="Input"
  />
</DocumentFragment>
`;
